ladybug/_datacollectionbase.py

```diff
@@ -0,0 +1,517 @@
+            .replace("~", "not").replace("<<", "in").replace("$", "is")
+
+    def _filter_by_statement(self, statement):
+        """Filter the data collection based on a conditional statement."""
+        self.__class__._check_conditional_statement(statement, 1)
+        statement = statement.replace('a', 'd')
+        _filt_values, _filt_datetimes = [], []
+        for i, d in enumerate(self._values):
+            if eval(statement, {'d': d}):
+                _filt_values.append(d)
+                _filt_datetimes.append(self.datetimes[i])
+        return _filt_values, _filt_datetimes
+
+    def _filter_by_pattern(self, pattern):
+        """Filter the Filter the Data Collection based on a list of booleans."""
+        try:
+            _len = len(pattern)
+        except TypeError:
+            raise TypeError("pattern is not a list of Booleans. Got {}".format(
+                type(pattern)))
+        _filt_values = [d for i, d in enumerate(self._values) if pattern[i % _len]]
+        _filt_datetimes = [d for i, d in enumerate(self.datetimes) if pattern[i % _len]]
+        return _filt_values, _filt_datetimes
+
+    def _is_in_data_type_range(self, raise_exception=True):
+        """Check if the Data Collection values are in permissable ranges for the data_type.
+
+        If this method returns False, the Data Collection's data is
+        physically or mathematically impossible for the data_type."""
+        return self._header.data_type.is_in_range(
+            self._values, self._header.unit, raise_exception)
+
+    def _is_in_epw_range(self, raise_exception=True):
+        """Check if Data Collection values are in permissable ranges for EPW files."""
+        return self._header.data_type.is_in_range_epw(
+            self._values, self._header.unit, raise_exception)
+
+    def _percentile(self, values, percent, key=lambda x: x):
+        """Find the percentile of a list of values.
+
+        Args:
+            values: A list of values for which percentiles are desired
+            percent: A float value from 0 to 100 representing the requested percentile.
+            key: optional key function to compute value from each element of N.
+
+        Return:
+            The percentile of the values
+        """
+        vals = sorted(values)
+        k = (len(vals) - 1) * (percent / 100)
+        f = math.floor(k)
+        c = math.ceil(k)
+        if f == c:
+            return key(vals[int(k)])
+        d0 = key(vals[int(f)]) * (c-k)
+        d1 = key(vals[int(c)]) * (k-f)
+        return d0+d1
+
+    def __len__(self):
+        return len(self._values)
+
+    def __getitem__(self, key):
+        return self._values[key]
+
+    def __setitem__(self, key, value):
+        self._values[key] = value
+
+    def __delitem__(self, key):
+        del self._values[key]
+        del self._datetimes[key]
+
+    def __iter__(self):
+        return iter(self._values)
+
+    def __reversed__(self):
```

Comment:
I really didn't think of it that way but it makes sense.  The one thing that makes me a bit uneasy about doing that is that, currently, none of the methods that could be applied to annual hourly data would result in the datetimes getting out of chronological order.  While everything I have implemented now would still work even with datetimes out of order, I am not sure that I want to introduce things that will deliberately make the datetimes out of order since I can potentially see this as useful for some workflows.  For example, if there are "holes" in a discontinuous data set over which I want to interpolate (very common with data in the NOAA database), knowing that the datetimes are chronological will avoid the need to sort the datetimes.  Accordingly, I think it is better practice just to take this method out for now.